core/src/main/java/org/elasticsearch/index/query/GeoBoundingBoxQueryBuilder.java

```diff
@@ -97,94 +155,286 @@ public GeoBoundingBoxQueryBuilder bottomRight(String geohash) {
+    public boolean ignoreMalformed() {
+        return ignoreMalformed;
+    }
+
     /**
      * Sets the type of executing of the geo bounding box. Can be either `memory` or `indexed`. Defaults
      * to `memory`.
      */
-    public GeoBoundingBoxQueryBuilder type(String type) {
+    public GeoBoundingBoxQueryBuilder type(GeoExecType type) {
+        if (type == null) {
+            throw new IllegalArgumentException("Type is not allowed to be null.");
+        }
         this.type = type;
         return this;
     }
 
+    /**
+     * For BWC: Parse type from type name.
+     * */
+    public GeoBoundingBoxQueryBuilder type(String type) {
+        this.type = GeoExecType.fromString(type);
+        return this;
+    }
+    /** Returns the execution type of the geo bounding box.*/
+    public GeoExecType type() {
+        return type;
+    }
+
+    /** Returns the name of the field to base the bounding box computation on. */
+    public String fieldName() {
+        return this.fieldName;
+    }
+
     @Override
-    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
-        // check values
-        if(Double.isNaN(box[TOP])) {
-            throw new IllegalArgumentException("geo_bounding_box requires top latitude to be set");
-        } else if(Double.isNaN(box[BOTTOM])) {
-            throw new IllegalArgumentException("geo_bounding_box requires bottom latitude to be set");
-        } else if(Double.isNaN(box[RIGHT])) {
-            throw new IllegalArgumentException("geo_bounding_box requires right longitude to be set");
-        } else if(Double.isNaN(box[LEFT])) {
-            throw new IllegalArgumentException("geo_bounding_box requires left longitude to be set");
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (!ignoreMalformed) {
+            if (Numbers.isValidDouble(topLeft.getLat()) == false) {
+                validationException = addValidationError("top latitude is invalid: " + topLeft.getLat(), validationException);
+            }
+            if (Numbers.isValidDouble(topLeft.getLon()) == false) {
+                validationException = addValidationError("left longitude is invalid: " + topLeft.getLon(), validationException);
+            }
+            if (Numbers.isValidDouble(bottomRight.getLat()) == false) {
+                validationException = addValidationError("bottom latitude is invalid: " + bottomRight.getLat(), validationException);
+            }
+            if (Numbers.isValidDouble(bottomRight.getLon()) == false) {
+                validationException = addValidationError("right longitude is invalid: " + bottomRight.getLon(), validationException);
+            }
+
+            if (validationException == null) {
+                // all corners are valid after above checks - make sure they are in the right relation
+                if (topLeft.getLat() < bottomRight.getLat()) {
```

Comment:
can you rearrange this such that we can check this once it's set. ie instead of setting these values individually we only provide methods to set both?